src/main/java/org/sonatype/nexus/repository/composer/internal/ComposerFormatAttributesExtractor.java

```diff
@@ -0,0 +1,241 @@
+    try (InputStream is = tempBlob.getBlob().getInputStream()) {
+      try (ArchiveInputStream ais = archiveStreamFactory.createArchiveInputStream(ArchiveStreamFactory.ZIP, is)) {
+        ArchiveEntry entry = ais.getNextEntry();
+        while (entry != null) {
+          if (processEntry(ais, entry, formatAttributes)) {
+            return;
+          }
+          entry = ais.getNextEntry();
+        }
+      }
+    }
+    catch (ArchiveException e) {
+      throw new IOException("Error reading from archive", e);
+    }
+  }
+
+  /**
+   * Processes a single entry in the archive. If the entry is the composer.json then the attributes will be extracted.
+   * If not, the entry is skipped.
+   */
+  private boolean processEntry(final ArchiveInputStream stream,
+                               final ArchiveEntry entry,
+                               final NestedAttributesMap formatAttributes) throws IOException
+  {
+    String name = entry.getName();
+    int filenameIndex = name.indexOf("/composer.json");
+    int separatorIndex = name.indexOf("/");
+    if (filenameIndex >= 0 && filenameIndex == separatorIndex) {
+      Map<String, Object> contents = mapper.readValue(stream, typeReference);
+      extractStrings(contents, formatAttributes, STRINGS_MAPPING);
+      extractAuthors(contents, formatAttributes);
+      extractSupport(contents, formatAttributes);
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * Extracts zero or more string-only fields from the source map into the destination attribute map. If a collection
+   * is encountered, any string items within the collection are added to a list and stored as a collection of strings.
+   */
+  @VisibleForTesting
+  void extractStrings(final Map<String, Object> source,
+                      final NestedAttributesMap destination,
+                      final Map<String, String> mappings)
+  {
+    for (Map.Entry<String, String> mapping : mappings.entrySet()) {
+      Object sourceValue = source.get(mapping.getKey());
+      if (sourceValue instanceof String) {
+        destination.set(mapping.getValue(), sourceValue);
+      }
+      else if (sourceValue instanceof Collection) {
+        List<String> entries = new ArrayList<>();
+        for (Object entryValue : (Collection) sourceValue) {
+          if (entryValue instanceof String) {
+            entries.add((String) entryValue);
+          }
+        }
+        if (!entries.isEmpty()) {
+          destination.set(mapping.getValue(), entries);
+        }
+      }
+    }
+  }
+
+  /**
+   * Extracts author contact information (except for the role) into a collection of strings.
+   */
+  @VisibleForTesting
+  void extractAuthors(final Map<String, Object> contents,
+                      final NestedAttributesMap formatAttributes)
+  {
+    Object sourceValue = contents.get(AUTHORS);
+    if (sourceValue instanceof Collection) {
```

Comment:
Trivial: Should we log if the author information could not be found in the composer file

Reply:
> Trivial: Should we log if the author information could not be found in the composer file

I'm not that worried about a missing piece of information like this since it's not something germane to the actual functioning of the implementation; we don't validate each piece of metadata for the other formats, though I concede in this case we could make an argument for doing so because of the existence of a defined schema for the JSON. When we get to the hosted impl there will be some fields that we do care about (we cannot proceed without them, i.e. the version field) in general if there's a missing field.